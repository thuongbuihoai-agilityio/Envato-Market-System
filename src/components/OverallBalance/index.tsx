import { memo, useCallback, useState } from 'react';
import isEqual from 'react-fast-compare';

// Components
import Chart from 'react-apexcharts';
import {
  Box,
  Flex,
  Heading,
  Text,
  theme,
  useColorModeValue,
} from '@chakra-ui/react';

// Icon
import { Arrow } from '@app/components/Icons';

// Constants
import {
  OVERALL_BALANCE_COLORS,
  REVENUE_FLOW_OPTIONS,
  REVENUE_FLOW_STATUS,
} from '@app/constants';

// Types
import { IRevenueFlow } from '@app/interfaces';

// Components
import Select, { TOption } from '@app/components/common/Select';

// Utils
import { formatDecimalNumber } from '@app/utils';

type TOverallData = Omit<IRevenueFlow, 'pending'>[];

interface OverallBalanceProps {
  total: number;
  growth: number;
  data: TOverallData;
}

const OverallBalanceComponent = ({
  data,
  total,
  growth,
}: OverallBalanceProps) => {
  const [chartData, setChartData] = useState<TOverallData>(data);

  const colorFill = useColorModeValue(
    theme.colors.gray[800],
    theme.colors.white,
  );

  const renderTitle = useCallback(
    ({ label }: TOption) => (
      <Flex alignItems="center">
        <Text>{label}</Text>
        <Arrow color={colorFill} />
      </Flex>
    ),
    [colorFill],
  );

  const handleChangeSelect = useCallback(
    ({ value }: TOption) => {
      const temp: TOverallData = [...data];
      const result: Record<string, TOverallData> = {
        'Jan,Jun': temp.slice(0, -6),
        'July,Dec': temp.slice(-6),
      };

      setChartData(result[value] || data);
    },
    [data],
  );

  return (
    <Box py={3} px={6} bg="background.component.primary" rounded="lg">
      <Flex
        py={4}
        px={5}
        borderBottom="1px"
        borderColor="border.primary"
        justifyContent="space-between"
      >
        <Box>
          <Text variant="textSm">Overall Balance</Text>
          <Flex align="center" gap={2}>
            <Heading variant="heading2Xl" as="h3">
              ${formatDecimalNumber(total)}
            </Heading>
            <Text color="primary.500">{growth}%</Text>
          </Flex>
        </Box>
        <Flex gap={7} display={{ base: 'none', lg: 'flex' }}>
          {REVENUE_FLOW_STATUS.slice(0, -1).map((item, index) => (
            <Flex key={item} gap={2} alignItems="center">
              <Box
                bgColor={OVERALL_BALANCE_COLORS[index]}
                w={3}
                height={3}
                rounded="50%"
              />
              <Text variant="textSm">{item}</Text>
            </Flex>
          ))}
        </Flex>
        <Box w={125} h="21px">
          <Select
            options={REVENUE_FLOW_OPTIONS}
            variant="no-border"
            renderTitle={renderTitle}
            onSelect={handleChangeSelect}
          />
        </Box>
      </Flex>
      <Chart
        options={{
          chart: {
            stacked: true,
            toolbar: {
              show: false,
            },
          },
          xaxis: {
            categories: chartData.map((item) => item.title),
            axisTicks: {
              show: false,
            },
            labels: {
              style: {
                colors: colorFill,
              },
            },
          },
          yaxis: {
            labels: {
              style: {
                colors: colorFill,
              },
            },
          },
          legend: {
            show: false,
          },
          colors: OVERALL_BALANCE_COLORS,
          dataLabels: {
            enabled: false,
          },
          tooltip: {
            custom: function ({ series, dataPointIndex }) {
              return `<div style="padding: 10px; background-color: #000; color: #FFF">
                <div>
                ${chartData[dataPointIndex].title}
                </div>
                <p>
                ${REVENUE_FLOW_STATUS[0]}: ${series[0][dataPointIndex]}
                </p>
                <p>
                ${REVENUE_FLOW_STATUS[1]}: ${series[1][dataPointIndex]}
                </p>
                </div>`;
            },
          },
        }}
        series={[
          {
            data: chartData.map(({ signed }) => signed),
          },
          {
            data: chartData.map(({ lost }) => lost),
          },
        ]}
        type="area"
        height="210"
      />
    </Box>
  );
};

const OverallBalance = memo(OverallBalanceComponent, isEqual);

export default OverallBalance;
