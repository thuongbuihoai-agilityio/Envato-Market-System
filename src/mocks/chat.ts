// Constants
import { AVATAR_POSITION, IMAGES } from '@app/constants';
import { MESSAGE_TIME } from '.';

// TODO: Update later
export const USER_CHATS = [
  {
    uid: 'user',
    isSend: false,
    isAudio: false,
    content: 'Hi, I need more information',
    avatarPosition: AVATAR_POSITION.BEFORE,
    avatar: IMAGES.CHAT_USER_AVATAR.url,
    localeTime: MESSAGE_TIME + 1000,
  },

  {
    uid: 'user',
    isSend: false,
    isAudio: false,
    content: 'Hello, I want to know more about your product',
    avatarPosition: AVATAR_POSITION.BEFORE,
    avatar: IMAGES.CHAT_USER_AVATAR.url,
    localeTime: MESSAGE_TIME + 3000,
  },

  {
    uid: 'user',
    isSend: false,
    isAudio: false,
    content: 'Hello, I want to know more about your product',
    avatarPosition: AVATAR_POSITION.BEFORE,
    avatar: IMAGES.CHAT_USER_AVATAR.url,
    localeTime: MESSAGE_TIME + 6000,
  },

  {
    uid: 'admin',
    isSend: true,
    isAudio: false,
    content: 'Hi, i am here for help',
    avatarPosition: AVATAR_POSITION.AFTER,
    avatar: IMAGES.CHAT_USER_AVATAR.url,
    localeTime: MESSAGE_TIME + 2000,
  },

  {
    uid: 'admin',
    isSend: true,
    isAudio: false,
    content: 'Sure, I can help you with that',
    avatarPosition: AVATAR_POSITION.BEFORE,
    avatar: IMAGES.CHAT_USER_AVATAR.url,
    localeTime: MESSAGE_TIME + 4000,
  },
  {
    uid: 'admin',
    isSend: true,
    isAudio: false,
    content: 'Sure, I can help you with that',
    avatarPosition: AVATAR_POSITION.BEFORE,
    avatar: IMAGES.CHAT_USER_AVATAR.url,
    localeTime: MESSAGE_TIME + 7000,
  },
];

export const MEMBER_CHATS = [
  { uid: 'admin', name: 'Admin' },
  { uid: 'user1', name: 'User 1' },
];
