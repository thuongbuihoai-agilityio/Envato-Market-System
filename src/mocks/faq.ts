export const FAQ_DATA = [
  {
    id: 1,
    question: 'What is the most important part of a dashboard?',
    answer:
      'In reality, the most important aspect of a great dashboard is the part that gets the least amount of attention: The underlying data. More than any other aspect, the data will make or break a dashboard.Within this definition, successful administration appears to rest on three basic skills, which we will call technical, human, and conceptual.Dashboards are business intelligence.',
  },
  {
    id: 2,
    question: 'What are the three types of dashboard?',
    answer:
      'In reality, the most important aspect of a great dashboard is the part that gets the least amount of attention: The underlying data. More than any other aspect, the data will make or break a dashboard.Within this definition, successful administration appears to rest on three basic skills, which we will call technical, human, and conceptual.Dashboards are business intelligence.',
  },
  {
    id: 3,
    question: 'What are examples of admin?',
    answer:
      'In reality, the most important aspect of a great dashboard is the part that gets the least amount of attention: The underlying data. More than any other aspect, the data will make or break a dashboard.Within this definition, successful administration appears to rest on three basic skills, which we will call technical, human, and conceptual.Dashboards are business intelligence.',
  },
  {
    id: 4,
    question: 'What are 5 benefits of dashboards?',
    answer:
      'In reality, the most important aspect of a great dashboard is the part that gets the least amount of attention: The underlying data. More than any other aspect, the data will make or break a dashboard.Within this definition, successful administration appears to rest on three basic skills, which we will call technical, human, and conceptual.Dashboards are business intelligence.',
  },
  {
    id: 5,
    question: 'What makes a good dashboard?',
    answer:
      'In reality, the most important aspect of a great dashboard is the part that gets the least amount of attention: The underlying data. More than any other aspect, the data will make or break a dashboard.Within this definition, successful administration appears to rest on three basic skills, which we will call technical, human, and conceptual.Dashboards are business intelligence.',
  },
  {
    id: 6,
    question: 'What are the 3 basic skills of administrator?',
    answer:
      'In reality, the most important aspect of a great dashboard is the part that gets the least amount of attention: The underlying data. More than any other aspect, the data will make or break a dashboard.Within this definition, successful administration appears to rest on three basic skills, which we will call technical, human, and conceptual.Dashboards are business intelligence.',
  },
];
