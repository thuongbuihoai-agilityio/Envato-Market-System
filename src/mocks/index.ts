export * from './spending';
export * from './select';
export * from './table';
export * from './sidebar';
