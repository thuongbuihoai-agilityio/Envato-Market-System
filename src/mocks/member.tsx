import { Image } from '@chakra-ui/react';

// Constants
import { IMAGES } from '@app/constants';

// Utils
import { MESSAGE_TIME_FORMAT } from '.';

export const MEMBER_CHAT = [
  {
    id: 'user1',
    avatar: 'https://example.com/avatar1.jpg',
    name: 'Kien Nguyen',
    status: 'Online',
    localeTime: MESSAGE_TIME_FORMAT,
    icon: (
      <Image
        w={5}
        h={5}
        src="icons/file.svg"
        alt={IMAGES.ATTACH.alt}
        fallbackSrc={IMAGES.FALLBACK.url}
      />
    ),
    statusColor: 'online',
  },
  {
    id: 'user2',
    avatar: 'https://example.com/avatar2.jpg',
    name: 'Huy Pham',
    status: 'Away',
    localeTime: MESSAGE_TIME_FORMAT,

    icon: (
      <Image
        w={5}
        h={5}
        src="icons/file.svg"
        alt={IMAGES.ATTACH.alt}
        fallbackSrc={IMAGES.FALLBACK.url}
      />
    ),
    statusColor: 'offline',
  },
];
