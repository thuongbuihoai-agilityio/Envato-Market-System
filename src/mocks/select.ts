export const SOCIAL_PLATFORM_OPTIONS = [
  {
    value: 'github',
    label: 'github',
  },
  {
    value: 'facebook',
    label: 'facebook',
  },
  {
    value: 'instagram',
    label: 'instagram',
  },
];
