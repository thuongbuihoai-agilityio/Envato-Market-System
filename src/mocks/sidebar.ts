// constants
import {
  HELP_ITEM_LIST,
  MENU_ITEM_LIST,
  OTHER_ITEM_LIST,
} from '@app/constants/sidebar';

export const EXPAND_SIDEBAR_MENU_LIST = [
  {
    id: 1,
    title: 'Menu',
    listItem: MENU_ITEM_LIST,
  },
  {
    id: 2,
    title: 'Help',
    listItem: HELP_ITEM_LIST,
  },
  {
    id: 3,
    title: 'Others',
    listItem: OTHER_ITEM_LIST,
  },
];
