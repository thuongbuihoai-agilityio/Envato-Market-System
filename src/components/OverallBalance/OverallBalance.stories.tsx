import { StoryObj, Meta } from '@storybook/react';

// Components
import OverallBalance from '@app/components/OverallBalance';

// Mocks
import { OVERALL_BALANCE_MOCK } from '@app/mocks';

const meta: Meta<typeof OverallBalance> = {
  title: 'Custom Components/OverallBalance',
  tags: ['autodocs'],
  component: OverallBalance,
  argTypes: {
    data: {
      description:
        'The data of the overall balance for 12 months includes: pending, signed and lost',
    },
    growth: {
      description: 'The growth amount of the overall balance',
    },
    total: {
      description: 'The total amount of the overall balance',
    },
  },
  parameters: {
    controls: {
      expanded: true,
    },
  },
};

export default meta;
type Story = StoryObj<typeof OverallBalance>;

export const Default: Story = {
  args: OVERALL_BALANCE_MOCK,
};
