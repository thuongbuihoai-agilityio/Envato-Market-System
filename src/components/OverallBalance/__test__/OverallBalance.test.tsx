import { render } from '@testing-library/react';

// Components
import OverallBalance from '..';

// Mock
import { OVERALL_BALANCE_MOCK } from '@app/mocks';

jest.mock('react-apexcharts', () => ({
  __esModule: true,
  default: () => <div />,
}));

describe('OverallBalance component', () => {
  it('renders correctly', () => {
    const { container } = render(<OverallBalance {...OVERALL_BALANCE_MOCK} />);

    expect(container).toMatchSnapshot();
  });
});
