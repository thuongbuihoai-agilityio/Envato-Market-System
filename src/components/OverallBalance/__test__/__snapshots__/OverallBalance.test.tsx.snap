// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OverallBalance component renders correctly 1`] = `
<div>
  <div
    class="css-p120lk"
  >
    <div
      class="css-1spmqp"
    >
      <div
        class="css-0"
      >
        <p
          class="chakra-text css-0"
        >
          Overall Balance
        </p>
        <div
          class="css-j2at52"
        >
          <h3
            class="chakra-heading css-0"
          >
            $
            48.574
          </h3>
          <p
            class="chakra-text css-ss2gmn"
          >
            20
            %
          </p>
        </div>
      </div>
      <div
        class="css-vd6ngq"
      >
        <div
          class="css-r2hvuq"
        >
          <div
            class="css-18ek59d"
          />
          <p
            class="chakra-text css-0"
          >
            Pending
          </p>
        </div>
        <div
          class="css-r2hvuq"
        >
          <div
            class="css-vdd5pp"
          />
          <p
            class="chakra-text css-0"
          >
            Signed
          </p>
        </div>
        <div
          class="css-r2hvuq"
        >
          <div
            class="css-y7nxfq"
          />
          <p
            class="chakra-text css-0"
          >
            Lost
          </p>
        </div>
      </div>
      <div
        class="css-19vtkro"
      >
        <button
          aria-controls="menu-list-:r1:"
          aria-expanded="false"
          aria-haspopup="menu"
          class="chakra-button chakra-menu__menu-button css-1rasy8u"
          id="menu-button-:r1:"
          type="button"
        >
          <span
            class="css-xl71ch"
          >
            <div
              class="css-k008qs"
            >
              <p
                class="chakra-text css-0"
              >
                Weekly
              </p>
              <svg
                fill="none"
                height="16"
                stroke="#1A202C"
                style="transform: rotate(undefined);"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M7 10L12 14L17 10"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </div>
          </span>
        </button>
        <div
          class="css-0"
          style="visibility: hidden; position: absolute; inset: 0 auto auto 0;"
        >
          <div
            aria-orientation="vertical"
            class="chakra-menu__menu-list css-ta0hp9"
            id="menu-list-:r1:"
            role="menu"
            style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8) translateZ(0);"
            tabindex="-1"
          >
            <button
              class="chakra-menu__menuitem css-1oz22gy"
              data-index="0"
              id="menu-list-:r1:-menuitem-:r2:"
              role="menuitem"
              tabindex="-1"
              type="button"
            >
              <p
                class="chakra-text css-1198d6z"
              >
                Weekly
              </p>
            </button>
            <button
              class="chakra-menu__menuitem css-1oz22gy"
              data-index="1"
              id="menu-list-:r1:-menuitem-:r3:"
              role="menuitem"
              tabindex="-1"
              type="button"
            >
              <p
                class="chakra-text css-1198d6z"
              >
                monthly    
              </p>
            </button>
            <button
              class="chakra-menu__menuitem css-1oz22gy"
              data-index="2"
              id="menu-list-:r1:-menuitem-:r4:"
              role="menuitem"
              tabindex="-1"
              type="button"
            >
              <p
                class="chakra-text css-1198d6z"
              >
                Yearly
              </p>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div />
  </div>
</div>
`;

exports[`OverallBalance component renders with is loading is true 1`] = `
<div>
  <div
    class="chakra-skeleton css-76u8cq"
  />
</div>
`;
